### ISSUE TITLE
Support checking return statements against `@return array{0:T,1:T2}` types?

### ISSUE THREAD

**Issue by TysonAndre** [vimeo/psalm#398]
Expected: no error
Observed:  `The type 'array<int, stdClass|int>' does not match the declared return type 'array{0:int, 1:stdClass}' for test`

E.g. if it's an inline return statement, then check if the return type has array fields before deciding how the array type would be computed

It seems like this is specific to numeric keys, string keys work as expected.

Use case: Functions that return tuples of values (e.g. error code + result). Fairly uncommon.

```php
<?php

/**
 * @return array{0:?int,1:stdClass}
 */
function test() {
    return [44, new stdClass()];  // Unexpected InvalidReturnStatement
}
```

**Comment by weirdan** [comment]
Also see #335

**Comment by muglug** [comment]
What's necessary here is to fully transition logic designed for string-offset arrays to integer-offset arrays. Right now the current support for int-offset keys is a bit of a hack, and needs to be more robust.

**Comment by weirdan** [comment]
Is there some syntax to make all of those keys required? Currently, where I pass two-element array `[1,1]` to function requiring `array{0:int,5:string}` parameter Psalm thinks it's all fine.

**Comment by muglug** [comment]
@weirdan Upgrade Psalm I think? I fixed this yesterday.

This is now an error:
```php
/** @param array{0:int, 5:string} $arr */
function takesArr(array $arr) : void {}
takesArr([1, 1]);
```

**Comment by weirdan** [comment]
@muglug right, that works. There's something weird going on with `array_reduce` then, so I going to report that separately.
